target/linux/realtek/files-5.10/drivers/net/ethernet/rtl838x_eth.c

```diff
@@ -839,17 +839,37 @@ static void rtnc_93xx_header_vlan_set(struct rtnc_hdr *h, int vlan)
 	h->cpu_tag[3] |= (vlan & 0xff) << 8;
 }
 
-void rtnc_83xx_update_cntr(int r, int released)
+/*
+ * On all SOCs, the DMA_IF_RX_RING_CNTR track the fill level of the rings.
+ * Writing x into these registers substracts x from its content.
+ * When the DMA_IF_RX_RING_SIZE is not zero and the content reaches the ring
+ * size, the SOC no longer adds packets to a receive queue.
+ */
+
+void rtnc_838x_update_cntr(int r, int released)
 {
-	/* This feature is not available on RTL838x/RTL839x SoCs */
+	int pos = r * 4;
+	u32 reg = rtnc_838x_dma_if_rx_ring_cntr(r);
+
+	/*
+	 * The RTL838X counter modifications are not atomic. A decrement
+	 * from the CPU might get lost when new packets arrive and the counter
+	 * is increased in the same moment from the SOC. As software buffers
+	 * are much larger than the maximum possible value of 15 it is no
+	 * problem to clear the counter.
+	 */
+	sw_w32(sw_r32(reg) & (0xf << pos), reg);
+}
+
+void rtnc_839x_update_cntr(int r, int released)
+{
+	/*
+	 * The RTL839X is the only SOC that will not crash when it finds
+	 * a CPU owned ring buffer. No ring size limit is set and there is
+	 * no need to update the counters accordingly.
+	 */
```

Comment:
Would it be bad however to HAVE the function here? We have the registers, all other SoC's update the buffers, doesn't hurt to keep it does it?